### DISCUSSION TITLE
Multiple views per component

### DISCUSSION THREAD

**Discussion by tanssinet** [django-components/django-components#449] [Ideas]
For a few days, I have been working on a component that includes autocomplete. It may be a generic autocomplete component in the future, but currently it is related to the business logic.

I have been iimplementing it with django components, Alpine.js, HTMX and Bootstrap 5.

Currently, there are two components, "composition_search" and "composition_list_editor". The former is actually not a component in itself, but provides an HTMX endpoint to the latter, search results as &lt;div&gt; elements. So these two are tightly coupled.

I started thinking that perhaps they are so tightly coupled that they conceptually actually belong to the same component. The reason they are seperate components is that django components does not support having multiple different kind of endpoints per component, because there is view per component relationshp.

If you think about it, views are a way to express operations as endpoints, and it really makes sense that there can be many operations for a single component, especially in the HTMX world.

Currently, this is not a big issue for me, but just wanted to get the idea out there.

**Comment by dylanjcastillo** [comment]
This about [template fragments](https://htmx.org/essays/template-fragments/), right? I've been thinking about it but don't have a clear view now of how that'd work in django-components.

**Reply by mikko-ahonen** [reply]
@dylanjcastillo well it may be related, but I am coming to this from a different angle.

Perhaps "more than one view per compenent" is actually not that good title for what I mean.

What I would like to have is multiple operations (method calls) on a component. Now we only have few -- delete, get, post, and they do not necessarily correspond to the business  logic. Currently I encode the operation in the post data or URL parameters. It would be more clear if this would be handled by the django-component framework.

Simplest, although slightly contrived example: I have a component for editing of name of a model. One operation is opening the name for editing, and another is saving it. Now the post handles both of these, implicitly from the parameters. I would like the operation to be explicit. So I would like to have methods "start_editing", "save" and perhaps "cancel" that would be bound to different URLs.

**Reply by dylanjcastillo** [reply]
Thank you @mikko-ahonen. That was very useful.

Just to see if I understand what you mean, do you currently have something like this (maybe using query strings and with all the business logic inside `post`)?

**urls.py**
```python
urlpatterns = [
    ...
    path("multi_methods/<str:operation>/<int:id>", MultiMethods.as_view()),
    ...
]
```

**example_component.py**
```python
from django_components import component

@component.register("example")
class Example(component.Component):
    template = """<something></something>"""

    def get_context_data(self, id, **kwargs):
         ...

    def save(self, request, id, *args, **kwargs):
          ...

    def edit(self, request, id, *args, **kwargs):
          ...

    def post(self, request, operation, id, *args, **kwargs):
        if operation == "save":
            return self.save(request, id, *args, **kwargs)
        elif method == "edit":
            return self.edit(request, id, *args, **kwargs)
        else:
            pass
        return self.render_to_response(context)
```

And would want something that adds a bit more "magic", so that you don't explicitly need to use the post method to dispatch to the other methods?

**Reply by tanssinet** [reply]
Seems like I am using multiple accounts here, sorry for the confusiom. Same person anyway.

Yes, exactly. I feel like I am implementing a URL dispatcher inside each component.

**Reply by EmilStenstrom** [reply]
Personally I think this is something we shouldn't do as part of django-components. I think you could do it yourself though, just by overriding the Component class with your own logic and inheriting from that one instead. Makes sense?

**Reply by mikko-ahonen** [reply]
I guess it comes down to the question "what is a component?" and your vision for the library.

If you look at it technically, just as a set of js + python + html + css, then the current situation is fine, especially if you think that the "hardcore HTMX" way of development is not that important. 

However if we look at components from a logical perspective, I think one key element of a component is a support for multiple logical operations beyond the hard-coded ones (PUT, POST, GET, DELETE).  Another key aspect is that it encapsulates some reusable behavior.

For your proposed solution where there is some magic to route the request to a certain method, inheritance from Component makes sense.

I was thinking that there mightj be multiple alterative approaches to achieve the same functionality, such as multiple views per component.

In Django, there seems to be in practice an implicit assumptiion that a view and the corresponding URL are tightly coupled. If we also make an implicit assumption that one component is one view, in practice it seems to imply for the HTMX development that one logical component actually consists of multiple technical django-components.

**Reply by EmilStenstrom** [reply]
@mikko-ahonen Some other considerations:
* How much of what you are talking about is application specific, and how much of it is "base" functionality that should be part of django-components?
* What mental model do we expose to users of the library, both direct (people choosing the library and implementing it) and indirect (people reading the code that someone else implemented)?

I'm definitely open to discussing this further. For instance, what would the API be?

There are a couple of "framework" libraries out there, that are doing what you propose:

https://www.tetraframework.com/
```html
<button class="btn btn-primary" @click="add_todo(title)>
```
```python
@default.register
class ToDoList(Component):
    @public
    def add_todo(self, title):
        ...
```

https://www.django-unicorn.com/
```html
<form unicorn:submit.prevent="add">
```
```python
class TodoView(UnicornView):
    def add(self):
        ....
```

**Reply by mikko-ahonen** [reply]
I tried Unicorn, but got stuck with some "framework" problem. I have not tried Tetra, but it looks more promising as it is using Alpine. I prefer the library approach of Django components, though.

> How much of what you are talking about is application specific, and how much of it is "base" functionality that should be part of django-components?

I think it is core functionality of components, at least in how I see what components are.

>What mental model do we expose to users of the library, both direct (people choosing the library and implementing it) and indirect (people reading the code that someone else implemented)?

The mental model of the user and the API depends on the approach.

If we take the "component can be exposed as multiple views" approach, then I guess API would be some method decorators to define mapping between each  view and the methods, and adding view name to parameter for the as_view()

in components/foobar/foobar.py:
```
  @component.register("foobar")
  class Foobar(component.Component):
     ...
     @post_for_view("save")
     def save(...):
     ....
     @get_for_view("edit")
     def edit(...):
     ...
```

in components/urls.py:
```
    ...
    path("foobar/save", Foobar.as_view("save"), name='foobar-save'),
    path("foobar/edit", Foobar.as_view("edit"), name='foobar-edit'),
    ...
```

For the "component functions can be autowired to URL keyword" approach, then I guess the API would be subclass of Component + some code to expose methods as a view.

in components/foobar/foobar.py:
```
  @component.register("foobar")
  class Foobar(component.MethodViewsComponent):
     ...
     def save(...):
     ....
     def edit(...):
     ...
```

in components/urls.py:
```
    ...
    path("foobar/save", Foobar.methods_as_view(post="save"), name='foobar-save'),
    path("foobar/edit", Foobar.methods_as_view(get="edit"), name='foobar-edit'),
    ...
```
